crates/loam-cli/src/commands/dev/mod.rs

```diff
@@ -0,0 +1,219 @@
+use clap::Parser;
+use notify::{self, RecursiveMode, Watcher as NotifyWatcher};
+use std::{
+    env, fs,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+use tokio::sync::mpsc;
+use tokio::sync::Mutex;
+use tokio::time;
+
+use crate::commands::build;
+
+use super::build::clients::LoamEnv;
+
+enum Message {
+    FileChanged,
+}
+
+#[derive(Parser, Debug, Clone)]
+#[group(skip)]
+pub struct Cmd {
+    #[command(flatten)]
+    pub build_cmd: build::Cmd,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Watcher(#[from] notify::Error),
+    #[error(transparent)]
+    Build(#[from] build::Error),
+    #[error("IO error: {0}")]
+    Io(#[from] std::io::Error),
+}
+
+fn canonicalize_path(path: &Path) -> PathBuf {
+    if path.as_os_str().is_empty() {
+        env::current_dir().unwrap_or_else(|_| PathBuf::from("."))
+    } else if path.components().count() == 1 {
+        // Path is a single component, assuming it's a filename
+        env::current_dir()
+            .unwrap_or_else(|_| PathBuf::from("."))
+            .join(path)
+    } else {
+        fs::canonicalize(path).unwrap_or_else(|_| path.to_path_buf())
+    }
+}
+
+#[derive(Clone)]
+struct Watcher {
+    root_env: Arc<PathBuf>,
+    packages: Arc<Vec<PathBuf>>,
+}
+
+impl Watcher {
+    pub fn new(root_env: &Path, packages: &[PathBuf]) -> Self {
+        Self {
+            root_env: Arc::new(canonicalize_path(root_env)),
+            packages: Arc::new(packages.iter().map(|p| canonicalize_path(p)).collect()),
+        }
+    }
+
+    pub fn is_watched(&self, path: &Path) -> bool {
+        let path = canonicalize_path(path);
+        self.packages.iter().any(|p| path.starts_with(p))
+    }
+
+    pub fn is_env_toml(&self, path: &Path) -> bool {
+        canonicalize_path(path) == *self.root_env
+    }
+}
+
+fn is_temporary_file(path: &Path) -> bool {
+    const IGNORED_EXTENSIONS: &[&str] = &["tmp", "swp", "swo"];
+    let file_name = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
```

Comment:
The default af `""` seems strange. The reason you need to convert to a string is that a `str` is UTF-8 whereas  a Path uses an OsStr because some OS's can have path names which aren't safely converted to UTF-8. So it's trying to convert to a string and returning Option of whether it's valid UTF-8. We could consider using [camino](https://docs.rs/camino/latest/camino) which is a crate that enforces paths be UTF-8 (which for most rust projects is definitely the case). I've been wanting to try it out because then you can go between a Path and `str` without any headaches.

But for now I'd use an `expect` and we could use the other crate in a follow up PR.  I think it would clean up the whole repo.

Reply:
Okay, set it to use `expect` for now.